chalk-derive/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use syn::{parse_macro_input, Data, DeriveInput, GenericParam, Ident, TypeParamBound};
+
+#[proc_macro_derive(Fold, attributes(has_type_family))]
+pub fn derive_fold(item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as DeriveInput);
+
+    let name = input.ident;
+    let (_, ty_generics, where_clause_ref) = input.generics.split_for_impl();
+    let body = derive_fold_body(input.data);
+
+    // Allow custom Fold parameter using the `#[has_type_family(Param)]` attribute
+    let mut generics = if let Some(attr) = input
+        .attrs
+        .iter()
+        .find(|a| a.path.is_ident("has_type_family"))
+    {
+        let arg = attr
+            .parse_args::<proc_macro2::TokenStream>()
+            .expect("Expected has_type_family argument");
+        quote! { < #arg > } // TODO extend instead of overriding?
+    } else {
+        ty_generics.to_token_stream()
+    };
+
+    let mut impl_generics = input.generics.clone();
+    let mut where_clause = where_clause_ref.cloned();
+
+    // if any impl_generics G : TypeFamily:
+    // #impl_generics += TF: TypeFamily
+    // #generics = TF
+    // #where_clause += G: HasTypeFamily<TypeFamily = TF> + Fold<TF, Result = G>
+    if let Some(param) = input
+        .generics
+        .params
+        .iter()
+        .find_map(|p| is_type_family_param(p))
+    {
+        impl_generics.params.push(GenericParam::Type(
+            syn::parse(quote! { TF: TypeFamily }.into()).unwrap(),
+        ));
+        generics = quote! { <TF> };
```

Comment:
I think this should be

```rust
quote! { <<#param as HasTypeFamily>::TypeFamily> }
```

Reply:
in that case, we don't need to introduce a new `TF` type parameter or anything

Reply:
Good call, that also removed half of the where-clause.